### PR TITLE
feat: added web-first assertions

### LINKED ISSUE #988: Web Assertions for Python binding
https://github.com/microsoft/playwright/issues/8365 for Python

### PR DESCRIPTION
**Pull request by mxschmitt**
Fixes #988

### CODE REVIEW

File: playwright/_impl/_assertions.py [hunk 1]
```diff
@@ -0,0 +1,539 @@
...
+            message = message.replace("expected to", "expected not to")
+        if "useInnerText" in expect_options and expect_options["useInnerText"] is None:
+            del expect_options["useInnerText"]
+        result = await self._actual_locator._expect(expression, expect_options)
+        if result["matches"] == self._is_not:
+            log = "".join(result.get("log", ""))
+            if log.strip():
+                log = "\nCall log:\n" + log
+            if expected is None:
+                raise AssertionError(f"{message} {log}")
```
Comment (dgozman) [inline playwright/_impl/_assertions.py:50]: What's the difference between this and next lines?

File: playwright/_impl/_assertions.py [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+        self, title_or_reg_exp: Union[Pattern, str], timeout: float = None
+    ) -> None:
+        expected_values = to_expected_text_values(
+            [title_or_reg_exp], normalize_white_space=True
+        )
+        await self._expect_impl(
+            "to.have.title",
+            FrameExpectOptions(expectedText=expected_values, timeout=timeout),
+            title_or_reg_exp,
+            "Page title expected to be",
```
Comment (dgozman) [inline playwright/_impl/_assertions.py:73]: Could you please show an example of how the error looks like?

File: playwright/_impl/_assertions.py [hunk 3]
```diff
@@ -0,0 +1,539 @@
...
+            raise AssertionError(f"{message} {log}")
+
+
+class PageAssertions(AssertionsBase):
+    def __init__(self, page: Page, is_not: bool = False) -> None:
+        super().__init__(page.locator(":root"), is_not)
+        self._actual_page = page
+
+    @property
+    def _not(self) -> "PageAssertions":
```
Comment (dgozman) [inline playwright/_impl/_assertions.py:60]: Is this a public property?

File: playwright/_impl/_assertions.py [hunk 4]
```diff
@@ -0,0 +1,539 @@
...
+            expected["regexFlags"] += "s"
+        if (pattern.flags & int(re.MULTILINE)) != 0:
+            expected["regexFlags"] += "m"
+        assert (
+            pattern.flags
+            & ~(
+                int(re.MULTILINE)
+                | int(re.IGNORECASE)
+                | int(re.DOTALL)
+                | int(re.UNICODE)
```
Comment (dgozman) [inline playwright/_impl/_assertions.py:514]: JS has "u" flag - should we pass it as well?

File: playwright/_impl/_page.py [hunk 5]
```diff
@@ -817,6 +818,10 @@ async def wait_for_function(
     def workers(self) -> List["Worker"]:
         return self._workers.copy()
 
+    @property
+    def request(self) -> "APIRequestContext":
```
Comment (dgozman) [inline playwright/_impl/_page.py:822]: Seems unrelated.

File: playwright/async_api/__init__.py [hunk 6]
```diff
@@ -76,7 +82,28 @@ def async_playwright() -> PlaywrightContextManager:
     return PlaywrightContextManager()
 
 
+@overload
+def expect(page_or_locator: Page) -> PageAssertions:
+    ...
```
Comment (dgozman) [inline playwright/async_api/__init__.py:87]: What is this? Does it mean "pass"? I am confused :)

File: tests/async/test_assertions.py [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+# limitations under the License.
+
+import re
+from datetime import datetime
+
+import pytest
+
+from playwright.async_api import Browser, Page, expect
+from tests.server import Server
+
```
Comment (dgozman) [inline tests/async/test_assertions.py:22]: This does not look like it covers all cases we have it ts, does it?

Comment (mxschmitt) [reply tests/async/test_assertions.py:22]: This at least gave me 100% test-coverage, do you have any specific tests which you think are relevant for Python?

File: playwright/async_api/__init__.py (hunk 6 above)
Comment (mxschmitt) [reply playwright/async_api/__init__.py:87]: This is how Python handles overloads. Since the return value when you pass a page is different compared when you pass a locator.

File: playwright/_impl/_page.py (hunk 5 above)
Comment (mxschmitt) [reply playwright/_impl/_page.py:822]: Since we roll in this PR, this got caught up by it.

File: playwright/_impl/_assertions.py (hunk 4 above)
Comment (mxschmitt) [reply playwright/_impl/_assertions.py:514]: We didn't do it in Java, thats why I decided to not doing it and in Python every regexp has /u by default.

File: playwright/_impl/_assertions.py (hunk 3 above)
Comment (mxschmitt) [reply playwright/_impl/_assertions.py:60]: It is not, its used as a helper to easily cal the not_* methods.

File: playwright/_impl/_assertions.py (hunk 2 above)
Comment (mxschmitt) [reply playwright/_impl/_assertions.py:73]: https://gist.github.com/mxschmitt/1af1ee657a4237fafb13588b0324ceb3

File: playwright/_impl/_assertions.py (hunk 1 above)
Comment (mxschmitt) [reply playwright/_impl/_assertions.py:50]: In the first one we could access the actual / received values to create a better error message: https://github.com/microsoft/playwright-java/blob/3b4d8dc955f62cca263b1c31c5791a9d9e5fe122/assertions/src/main/java/com/microsoft/playwright/impl/AssertionsBase.java#L66-L69

In this specific case we don't yet, let me dig into formatting.

**Comment by kumaraditya303** [comment]
It would be better to have `expect(...).not.to_have_text` rather than `expect(...).not_to_have_text`.

**Comment by mxschmitt** [comment]
> It would be better to have `expect(...).not.to_have_text` rather than `expect(...).not_to_have_text`.

Yes we also thought the same, but unfortunately `not` is a keyword in Python and thats why we can't use it as a method name. We thought about `not_` instead but it felt wrong to us.

**Comment by kumaraditya303** [comment]
How about `expect_not(...).to_have_text` which is very explicit and concise ?

**Review by dgozman — APPROVED** [review]
Looks good to me.
